ratis-netty/src/main/java/org/apache/ratis/netty/server/NettyServerStreamRpc.java

```diff
@@ -292,35 +297,128 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) {
+      try {
+        server.submitClientRequestAsync(info.getRequest()).thenApplyAsync(reply -> {
+          if (reply.isSuccess()) {
+            ByteBuffer buffer = ClientProtoUtils.toRaftClientReplyProto(reply).toByteString().asReadOnlyByteBuffer();
+            sendReplySuccess(request, buffer, -1, ctx);
+          } else {
+            sendReplyNotSuccess(request, ctx);
+          }
+          return null;
+        });
+      } catch (IOException e) {
+        sendReplyNotSuccess(request, ctx);
+      }
+      return null;
+    });
+  }
+
   private void read(ChannelHandlerContext ctx, DataStreamRequestByteBuf request) {
     LOG.debug("{}: read {}", this, request);
     final ByteBuf buf = request.slice();
-    final boolean isHeader = request.getType() == Type.STREAM_HEADER;
 
     final StreamInfo info;
     final CompletableFuture<Long> localWrite;
     final List<CompletableFuture<DataStreamReply>> remoteWrites = new ArrayList<>();
     final StreamMap.Key key = new StreamMap.Key(ctx.channel().id(), request.getStreamId());
-    if (isHeader) {
+    if (request.getType() == Type.STREAM_HEADER) {
       info = streams.computeIfAbsent(key, id -> newStreamInfo(buf));
       localWrite = CompletableFuture.completedFuture(0L);
       for (DataStreamOutput out : info.getDataStreamOutputs()) {
         remoteWrites.add(out.getHeaderFuture());
       }
-    } else {
+    } else if (request.getType() == Type.STREAM_DATA) {
       info = streams.get(key);
       localWrite = info.getStream().thenApply(stream -> writeTo(buf, stream));
       for (DataStreamOutput out : info.getDataStreamOutputs()) {
         remoteWrites.add(out.writeAsync(request.slice().nioBuffer()));
       }
+    } else if (request.getType() == Type.STREAM_CLOSE) {
+      info = streams.get(key);
+      localWrite = info.getStream().thenApplyAsync(stream -> {
+        try {
+          stream.getWritableByteChannel().close();
+          return 0L;
+        } catch (IOException e) {
+          throw new CompletionException("Failed to close " + stream, e);
+        }
+      });
+
+      for (DataStreamOutput out : info.getDataStreamOutputs()) {
+        remoteWrites.add(out.closeAsync());
+      }
+    } else {
+      // peer server start transaction
+      peerServerStartTransaction(streams.get(key), request, ctx);
+      return;
     }
 
     final AtomicReference<CompletableFuture<?>> previous = info.getPrevious();
     final CompletableFuture<?> current = previous.get()
         .thenCombineAsync(JavaUtils.allOf(remoteWrites), (u, v) -> null)
         .thenCombineAsync(localWrite, (v, bytesWritten) -> {
           buf.release();
-          sendReply(remoteWrites, request, bytesWritten, ctx);
+          if (request.getType() == Type.STREAM_HEADER || request.getType() == Type.STREAM_DATA) {
+            sendReply(remoteWrites, request, bytesWritten, ctx);
+          } else if (request.getType() == Type.STREAM_CLOSE) {
+            if (info.getDataStreamOutputs().size() > 0) {
```

Comment:
When info.getDataStreamOutputs().size() == 0, it should call submitClientRequestAsync (i.e. primaryServerStartTransaction).

Reply:
@szetszwo Sorry. When info.getDataStreamOutputs().size() == 0, it means this server is a peer, it only needs to response to primary server with the close stream result. Maybe it need not to call primaryServerStartTransaction.

Reply:
How about the group only has one server?  It seems that the primary server won't call submitClientRequestAsync if there are no other peers.

Reply:
@szetszwo Thanks for reminding it, you are right, so I think we need to mark which server is primary: https://github.com/apache/incubator-ratis/pull/237/commits/7c578d754385734b98e56cdd65705c06ee381b10